system/Database/Postgre/Builder.php

```diff
@@ -312,6 +312,131 @@ public function join(string $table, $cond, string $type = '', ?bool $escape = nu
+
+            if ($constraints === []) {
+                if ($this->db->DBDebug) {
+                    throw new DatabaseException('You must specify a constraint to match on for batch updates.'); // @codeCoverageIgnore
+                }
+
+                return ''; // @codeCoverageIgnore
+            }
+
+            $updateFields = $this->QBOptions['updateFields'] ??
+                $this->updateFields($keys, false, $constraints)->QBOptions['updateFields'] ??
+                [];
+
+            $alias = $this->QBOptions['alias'] ?? '_u';
+
+            $sql = 'UPDATE ' . $this->compileIgnore('update') . $table . "\n";
+
+            $sql .= "SET\n";
+
+            $that = $this;
+            $sql .= implode(
+                ",\n",
+                array_map(
+                    static fn ($key, $value) => $key . ($value instanceof RawSql ?
+                            ' = ' . $value :
+                            ' = ' . $alias . '.' . $that->cast($value, $that->getFieldType($table, $key))),
+                    array_keys($updateFields),
+                    $updateFields
+                )
+            ) . "\n";
+
+            $sql .= "FROM (\n{:_table_:}";
+
+            $sql .= ') ' . $alias . "\n";
+
+            $sql .= 'WHERE ' . implode(
+                ' AND ',
+                array_map(
+                    static function ($key, $value) use ($table, $alias, $that) {
+                        if ($value instanceof RawSql && is_string($key)) {
+                            return $table . '.' . $key . ' = ' . $value;
+                        }
+
+                        if ($value instanceof RawSql) {
+                            return $value;
+                        }
+
+                        return $table . '.' . $value . ' = ' . $alias . '.' . $that->cast($value, $that->getFieldType($table, $value));
+                    },
+                    array_keys($constraints),
+                    $constraints
+                )
+            );
+
+            $this->QBOptions['sql'] = $sql;
+        }
+
+        if (isset($this->QBOptions['setQueryAsData'])) {
+            $data = $this->QBOptions['setQueryAsData'];
+        } else {
+            $data = implode(
+                " UNION ALL\n",
+                array_map(
+                    static fn ($value) => 'SELECT ' . implode(', ', array_map(
+                        static fn ($key, $index) => $index . ' ' . $key,
+                        $keys,
+                        $value
+                    )),
+                    $values
+                )
+            ) . "\n";
+        }
+
+        return str_replace('{:_table_:}', $data, $sql);
+    }
+
+    /**
+     * Returns cast expression.
+     *
+     * @TODO move this to BaseBuilder in 4.5.0
+     *
+     * @param float|int|string $expression
+     */
+    private function cast($expression, ?string $type): string
+    {
+        return ($type === null) ? $expression : $expression . '::' . strtoupper($type);
```

Comment:
```suggestion
        return ($type === null) ? $expression : 'CAST(' . $expression . ' AS ' . strtoupper($type) . ')';
```
The `expression::TYPE` format is a shorthand format postgre supports. However Postgre and all other DBMS support the format `CAST(expression AS type)`. We should go ahead and use this format so that if this method is moved to BaseBuilder it will generate code compatible with all DBMS.

https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-cast/
https://docs.oracle.com/javadb/10.8.3.0/ref/rrefsqlj33562.html
https://dev.mysql.com/doc/refman/8.0/en/cast-functions.html#function_cast
https://learn.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver16
https://www.sqlite.org/lang_expr.html#castexpr

Reply:
Done.